### ISSUE TITLE
past_key_values cat out of model generate, output appear disorder

### ISSUE THREAD

**Issue by lzlwakeup** [huggingface/transformers#35684]
### System Info

- `transformers` version: 4.47.1
- Platform: Linux-5.15.0-60-generic-x86_64-with-glibc2.35
- Python version: 3.11.11
- Huggingface_hub version: 0.27.0
- Safetensors version: 0.4.5
- Accelerate version: 1.2.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.4.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA 4090-24G

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

When I use past_key_values for generate, I have a problem. I want to cat prompt kv cache with current chat kv cache for limit gpu mem, but its fasle, this can be output, but not a right order sentence. 
 
code be like:

#prompt_kv_cache can be used
past_key_values = None
output = model.generate(input_ids,  past_key_values=past_key_values)
past_key_values = output.past_key_values

...some chat count 

last_past = []
for layer_past in past_key_values:
        past_keys, past_values = layer_past
        trimmed_past_keys =  past_keys[:, :, -len_kv:, :]
        trimmed_past_values =  past_values[:, :, -len_kv:, :]
        last_past.append((trimmed_past_keys, trimmed_past_values))
_past_all = []
    for p1_layer, p2_layer in zip(prompt_kv_cache, tuple(last_past)):     #prompt_kv_cache can be used for init, this true
        _past_all.append(tuple(torch.cat((p1, p2), dim=-2) for p1, p2 in zip(p1_layer, p2_layer)))
output = model.generate(input_ids,  past_key_values=tuple(_past_all))

Now, output.text is no-order.
For example, last input is "i want to eat"
output text maybe generate like "you \n\n need need some some some \n\n food food\n" or "you \n\n food \n need need some some some \n\n food food\n"

what wrong with past_key_values cat?

### Expected behavior

I want to know, how to cat a early kv cache with current kv cache
